### ISSUE TITLE
Early EOF in request body causes servlet to run multiple times

### ISSUE THREAD

**Issue by AgentME** [eclipse/jetty.project#2712]
I was curious how Jetty reacted to POST requests where the connection terminates early while the client is sending the request body (because of shoddy wifi, proxies going down, etc), so I decided to test it. I hoped that Jetty would block such connections from getting to my servlets or throw an error when the body is parsed (like when request.getParameter("...") is called), so that my servlets don't try to process corrupt incomplete data from users.

I got some very surprising results in the latest version of Jetty (9.4.11.v20180605) that seems like it must be a bug.

Here's the example program I'm testing with. It's in Kotlin. Sorry if anyone isn't familiar, but the example is hopefully pretty run-of-the-mill without many moving parts so it shouldn't be too hard to follow.

```kotlin
import org.eclipse.jetty.server.Server
import org.eclipse.jetty.server.ServerConnector
import org.eclipse.jetty.servlet.ServletHandler
import org.eclipse.jetty.servlet.ServletHolder
import org.eclipse.jetty.util.thread.QueuedThreadPool
import javax.servlet.http.HttpServlet
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

fun main(args: Array<String>) {
    val server = HttpServer()
    server.servletHandler.addServletWithMapping(ServletHolder(FooServlet()), "/foo")
    server.start()
    server.join()
}

class FooServlet : HttpServlet() {
    override fun doGet(req: HttpServletRequest, resp: HttpServletResponse) {
        resp.writer.println("received get")
    }

    override fun doPost(req: HttpServletRequest, resp: HttpServletResponse) {
        println("original status: ${resp.status}")
        println("bar parameter: ${req.getParameter("bar")}")
        resp.writer.println("received post ${req.getParameter("bar")}")
    }
}

class HttpServer {
    companion object {
        private const val PORT = 8080
        private const val MAX_SERVER_THREADS = 12
    }

    private val server: Server
    val servletHandler: ServletHandler // consumers should use methods like addServletWithMapping on this before starting

    init {
        val threadPool = QueuedThreadPool(MAX_SERVER_THREADS)
        threadPool.isDaemon = true
        server = Server(threadPool)
        val connector = ServerConnector(server)
        connector.port = PORT
        server.addConnector(connector)

        servletHandler = ServletHandler()
        server.handler = servletHandler
    }

    fun start() {
        server.start()
    }

    fun join() {
        server.join()
    }

    fun stop() {
        server.stop()
    }
}
```

The test setup so that someone can follow along. (The "badRequest" file was generated by using curl against `nc -l 8081 | tee badRequest`, and then I edited badRequest to have its Content-Length value be one higher than it should be.)

```
$ echo 'UE9TVCAvZm9vIEhUVFAvMS4xDQpIb3N0OiBsb2NhbGhvc3Q6ODA4MQ0KVXNlci1BZ2VudDogY3VybC83LjU0LjANCkFjY2VwdDogKi8qDQpDb250ZW50LUxlbmd0aDogMTINCkNvbnRlbnQtVHlwZTogYXBwbGljYXRpb24veC13d3ctZm9ybS11cmxlbmNvZGVkDQoNCmJhcj0xMjMmYz0x' | base64 --decode > badRequest
$ cat badRequest ; echo
POST /foo HTTP/1.1
Host: localhost:8081
User-Agent: curl/7.54.0
Accept: */*
Content-Length: 12
Content-Type: application/x-www-form-urlencoded

bar=123&c=1
```

## jetty 9.4.11.v20180605

```
$ nc localhost 8080 < badRequest
HTTP/1.1 500 Server Error
Content-Length: 18
Connection: close
Server: Jetty(9.4.11.v20180605)

received post 123
```

Server log:

```
[main] INFO org.eclipse.jetty.util.log - Logging initialized @1496ms to org.eclipse.jetty.util.log.Slf4jLog
[main] INFO org.eclipse.jetty.server.Server - jetty-9.4.11.v20180605; built: 2018-06-05T18:24:03.829Z; git: d5fc0523cfa96bfebfbda19606cad384d772f04c; jvm 1.8.0_131-b11
[main] INFO org.eclipse.jetty.server.AbstractConnector - Started ServerConnector@5158b42f{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
[main] INFO org.eclipse.jetty.server.Server - Started @2810ms
original status: 200
original status: 500
bar parameter: 123
```

It seems like calling request.getParameter("bar") throws an exception that Jetty silently catches, Jetty sets the response status to 500, and then Jetty re-runs the servlet for the same request. I assume this last part (re-running the servlet) is a bug because side effects in the servlet before the body is parsed are run twice and it implies a strange API (is every servlet supposed to check the response status first thing to detect if there's a parsing error?). I assume that after Jetty set the response status to 500, it was supposed to send an error message to the client and end the connection, but there happens to be a bug that causes it to loop back to the servlet.

## jetty 9.4.0.v20161208, 9.3.18.v20170406

I decided to test on some older version too to see what happened so I could figure out if the current behavior might be a recent bug (possibly caused by work mentioned in https://github.com/eclipse/jetty.project/issues/2562).

```
$ nc localhost 8080 < badRequest
HTTP/1.1 200 OK
Date: Wed, 11 Jul 2018 23:49:19 GMT
Content-Length: 18
Server: Jetty(9.4.0.v20161208)

received post 123
```

Server log:

```
[main] INFO org.eclipse.jetty.util.log - Logging initialized @1589ms to org.eclipse.jetty.util.log.Slf4jLog
[main] INFO org.eclipse.jetty.server.Server - jetty-9.4.0.v20161208
[main] INFO org.eclipse.jetty.server.AbstractConnector - Started ServerConnector@281a887{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
[main] INFO org.eclipse.jetty.server.Server - Started @2283ms
original status: 200
bar parameter: 123
[qtp399931359-18] WARN org.eclipse.jetty.server.Request - org.eclipse.jetty.io.EofException: Early EOF
```

(It seems to be non-deterministic where the WARN log messages appears relative to the original status and bar parameter lines, probably because the WARN goes through a Logger and stderr instead of just stdout.)

This old behavior seems unsurprising, but disappointing in that the early EOF exception doesn't seem to be exposed to the servlet at all, and the servlet goes on to process a corrupt request body.

**Comment by joakime** [comment]
Confirming that the badRequest demonstration has 11 bytes of data for an indicated `Content-Length: 12`, so it is indeed a bad request, that didn't send all of its data (early eof).

Early EOF is very common in the wild, browsers navigate away, mobile clients switch networks, etc ...
We don't log it anymore at the WARN level due to it being so prevalent / noisy in the wild.

Anyway, the heart of your question seems to be how to catch/response in code for this kind of situation?

Since the [`HttpServletRequest.getParameter(String)`](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getParameter-java.lang.String-) has no declared throwables, we have to rely on various runtime exceptions to communicate back to the user application.

Here's an example:

``` java
package org.eclipse.jetty.demo;

import java.io.IOException;

import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.eclipse.jetty.http.BadMessageException;
import org.eclipse.jetty.io.RuntimeIOException;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.DefaultHandler;
import org.eclipse.jetty.server.handler.HandlerList;
import org.eclipse.jetty.servlet.ServletContextHandler;

public class Demo
{
    public static void main(String[] args) throws Exception
    {
        Server server = new Server(8081);

        ServletContextHandler context = new ServletContextHandler();
        context.setContextPath("/");
        context.addServlet(FooServlet.class, "/foo");

        HandlerList handlers = new HandlerList();
        handlers.addHandler(context);
        handlers.addHandler(new DefaultHandler());

        server.setHandler(handlers);

        try
        {
            server.start();
            server.join();
        }
        catch (Throwable t)
        {
            t.printStackTrace();
        }
        finally
        {
            server.stop();
        }
    }

    public static class FooServlet extends HttpServlet
    {
        @Override
        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            resp.getWriter().println("received get");
        }

        @Override
        protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            System.out.printf("original status: %s%n", resp.getStatus());
            String barValue;

            try
            {
                barValue = req.getParameter("bar");
            }
            catch (BadMessageException e)
            {
                // we got the message, but there's a syntax / formatting / spec fault
                barValue = String.format("[BadMessageException: %s]", e.getMessage());
                e.printStackTrace(System.out);
            }
            catch (RuntimeIOException e)
            {
                // something wrong with how we read from the socket
                barValue = String.format("[RuntimeIOException: %s]", e.getMessage());
                e.printStackTrace(System.out);
            }
            catch (Throwable t)
            {
                // all other errors
                barValue = String.format("[Throwable: %s]", t.getMessage());
                t.printStackTrace(System.out);
            }

            System.out.printf("bar parameter: %s%n", barValue);
            resp.getWriter().printf("received post %s", barValue);
        }
    }
}
```

And a few examples ...

## Your badRequest

**Client Side**

``` shell
$ nc localhost 8081 < badRequest
HTTP/1.1 200 OK
Date: Thu, 12 Jul 2018 18:10:47 GMT
Content-Length: 80
Connection: close
Server: Jetty(9.4.11.v20180605)

received post [RuntimeIOException: org.eclipse.jetty.io.EofException: Early EOF]
```

**Server Side**

``` none
original status: 200
org.eclipse.jetty.io.RuntimeIOException: org.eclipse.jetty.io.EofException: Early EOF
	at org.eclipse.jetty.server.Request.extractFormParameters(Request.java:552)
	at org.eclipse.jetty.server.Request.extractContentParameters(Request.java:471)
	at org.eclipse.jetty.server.Request.getParameters(Request.java:386)
	at org.eclipse.jetty.server.Request.getParameter(Request.java:1025)
	at org.eclipse.jetty.demo.Demo$FooServlet.doPost(Demo.java:63)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:61)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.eclipse.jetty.io.EofException: Early EOF
	at org.eclipse.jetty.server.HttpInput$3.getError(HttpInput.java:1138)
	at org.eclipse.jetty.server.HttpInput$3.noContent(HttpInput.java:1126)
	at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:313)
	at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:256)
	at org.eclipse.jetty.util.UrlEncoded.decodeUtf8To(UrlEncoded.java:484)
	at org.eclipse.jetty.util.UrlEncoded.decodeTo(UrlEncoded.java:572)
	at org.eclipse.jetty.server.Request.extractFormParameters(Request.java:547)
	... 26 more
bar parameter: [RuntimeIOException: org.eclipse.jetty.io.EofException: Early EOF]
```

## Modified badRequest-syntax

**Client Side**

The request body contents contains a bad urlencoded sequence (`%foo;` is not valid)

``` shell
$ cat badRequest-syntax
POST /foo HTTP/1.1
Host: localhost:8081
User-Agent: curl/7.54.0
Accept: */*
Content-Length: 12
Content-Type: application/x-www-form-urlencoded

bar=%foo;123

$ nc localhost 8081 < badRequest-syntax
HTTP/1.1 200 OK
Date: Thu, 12 Jul 2018 18:12:48 GMT
Content-Length: 70
Server: Jetty(9.4.11.v20180605)

received post [BadMessageException: 400: Unable to parse form content]
```

**Server Side**

``` none
original status: 200
org.eclipse.jetty.http.BadMessageException: 400: Unable to parse form content
	at org.eclipse.jetty.server.Request.getParameters(Request.java:390)
	at org.eclipse.jetty.server.Request.getParameter(Request.java:1025)
	at org.eclipse.jetty.demo.Demo$FooServlet.doPost(Demo.java:63)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:61)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: Not valid encoding '%fo'
	at org.eclipse.jetty.util.UrlEncoded.decodeHexByte(UrlEncoded.java:889)
	at org.eclipse.jetty.util.UrlEncoded.decodeUtf8To(UrlEncoded.java:522)
	at org.eclipse.jetty.util.UrlEncoded.decodeTo(UrlEncoded.java:572)
	at org.eclipse.jetty.server.Request.extractFormParameters(Request.java:547)
	at org.eclipse.jetty.server.Request.extractContentParameters(Request.java:471)
	at org.eclipse.jetty.server.Request.getParameters(Request.java:386)
	... 24 more
bar parameter: [BadMessageException: 400: Unable to parse form content]
```

**Comment by AgentME** [comment]
Right, I am also interested in catching the error, so thanks for the great example!

But my other concern here is that Jetty's default handling of the error seems to be buggy and dangerous: the servlet is executed a second time for the same request if the error isn't caught.

**Comment by joakime** [comment]
If we use the original Servlet behavior like you reported, but add a `service()` method override to log all incoming requests, regardless of method type ...

``` java
    public static class OldFooServlet extends HttpServlet
    {
        @Override
        protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException
        {
            System.out.println("Entered Servlet: " + req);
            super.service(req, resp);
        }

        @Override
        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            resp.getWriter().println("received get");
        }

        @Override
        protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            System.out.printf("original status: %s%n", resp.getStatus());
            String barValue = req.getParameter("bar");
            System.out.printf("bar parameter: %s%n", barValue);
            resp.getWriter().printf("received post %s", barValue);
        }
    }
```

And make the same request you made before ...

``` none
$ nc localhost 8081 < badRequest
HTTP/1.1 500 Server Error
Cache-Control: must-revalidate,no-cache,no-store
Content-Type: text/html;charset=iso-8859-1
Content-Length: 2974
Connection: close
Server: Jetty(9.4.11.v20180605)

<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
<title>Error 500 Server Error</title>
</head>
<body><h2>HTTP ERROR 500</h2>
<p>Problem accessing /foo. Reason:
<pre>    Server Error</pre></p><h3>Caused by:</h3><pre>org.eclipse.jetty.io.EofException: Early EOF
        at org.eclipse.jetty.server.HttpInput$3.getError(HttpInput.java:1138)
        at org.eclipse.jetty.server.HttpInput$3.noContent(HttpInput.java:1126)
        at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:313)
        at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:256)
        at org.eclipse.jetty.util.UrlEncoded.decodeUtf8To(UrlEncoded.java:484)
        at org.eclipse.jetty.util.UrlEncoded.decodeTo(UrlEncoded.java:572)
        at org.eclipse.jetty.server.Request.extractFormParameters(Request.java:547)
        at org.eclipse.jetty.server.Request.extractContentParameters(Request.java:471)
        at org.eclipse.jetty.server.Request.getParameters(Request.java:386)
        at org.eclipse.jetty.server.Request.getParameter(Request.java:1025)
        at org.eclipse.jetty.demo.Demo$OldFooServlet.doPost(Demo.java:117)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
        at org.eclipse.jetty.demo.Demo$OldFooServlet.service(Demo.java:104)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1317)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1219)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
        at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:61)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handle(Server.java:531)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:762)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:680)
        at java.lang.Thread.run(Thread.java:748)
</pre>
<hr><a href="http://eclipse.org/jetty">Powered by Jetty:// 9.4.11.v20180605</a><hr/>

</body>
</html>
```

And then check the server ...

``` none
2018-07-12 13:31:24.196:INFO::main: Logging initialized @182ms to org.eclipse.jetty.util.log.StdErrLog
2018-07-12 13:31:24.383:INFO:oejs.Server:main: jetty-9.4.11.v20180605; built: 2018-06-05T18:24:03.829Z; git: d5fc0523cfa96bfebfbda19606cad384d772f04c; jvm 1.8.0_162-b12
2018-07-12 13:31:24.421:INFO:oejsh.ContextHandler:main: Started o.e.j.s.ServletContextHandler@71c7db30{/,null,AVAILABLE}
2018-07-12 13:31:24.580:INFO:oejs.AbstractConnector:main: Started ServerConnector@2812cbfa{HTTP/1.1,[http/1.1]}{0.0.0.0:8081}
2018-07-12 13:31:24.581:INFO:oejs.Server:main: Started @568ms
Entered Servlet: Request(POST //localhost:8081/foo)@1708f3b7
original status: 200
```

We see that this only results in 1 request to your Servlet.

**Comment by gregw** [comment]
@AgentME The second execution is probably an error dispatch?  Can you check the dispatcher type of the request on the first and second execution?

**Comment by AgentME** [comment]
@joakime It seems like the bug is specific to ServletHandler. (Your example used ServletContextHandler. Thanks for the tip for it! I didn't know about it and I'll be updating my projects for it.) Modifying your example like this reproduces the running-servlet-twice issue:

```java
package com.streak.emailmanager;

import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.eclipse.jetty.http.BadMessageException;
import org.eclipse.jetty.io.RuntimeIOException;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.DefaultHandler;
import org.eclipse.jetty.server.handler.HandlerList;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHandler;

public class Demo
{
	public static void main(String[] args) throws Exception
	{
		Server server = new Server(8081);

//		ServletContextHandler context = new ServletContextHandler();
//		context.setContextPath("/");
//		context.addServlet(FooServlet.class, "/foo");

		ServletHandler servletHandler = new ServletHandler();
		servletHandler.addServletWithMapping(FooServlet.class, "/foo");

//		HandlerList handlers = new HandlerList();
//		handlers.addHandler(context);
//		handlers.addHandler(new DefaultHandler());

//		server.setHandler(handlers);
		server.setHandler(servletHandler);

		try
		{
			server.start();
			server.join();
		}
		catch (Throwable t)
		{
			t.printStackTrace();
		}
		finally
		{
			server.stop();
		}
	}

	public static class FooServlet extends HttpServlet
	{
		@Override
		protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException
		{
			System.out.println("Entered Servlet: " + req);
			super.service(req, resp);
		}

		@Override
		protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
		{
			resp.getWriter().println("received get");
		}

		@Override
		protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
		{
			System.out.printf("original status: %s%n", resp.getStatus());
			String barValue;

//			try
//			{
				barValue = req.getParameter("bar");
//			}
//			catch (BadMessageException e)
//			{
//				// we got the message, but there's a syntax / formatting / spec fault
//				barValue = String.format("[BadMessageException: %s]", e.getMessage());
//				e.printStackTrace(System.out);
//			}
//			catch (RuntimeIOException e)
//			{
//				// something wrong with how we read from the socket
//				barValue = String.format("[RuntimeIOException: %s]", e.getMessage());
//				e.printStackTrace(System.out);
//			}
//			catch (Throwable t)
//			{
//				// all other errors
//				barValue = String.format("[Throwable: %s]", t.getMessage());
//				t.printStackTrace(System.out);
//			}

			System.out.printf("bar parameter: %s%n", barValue);
			resp.getWriter().printf("received post %s", barValue);
		}
	}
}
```

```
$ nc localhost 8081 < badRequest 
HTTP/1.1 500 Server Error
Content-Length: 17
Connection: close
Server: Jetty(9.4.11.v20180605)

received post 123
```

```
[main] INFO org.eclipse.jetty.util.log - Logging initialized @970ms to org.eclipse.jetty.util.log.Slf4jLog
[main] INFO org.eclipse.jetty.server.Server - jetty-9.4.11.v20180605; built: 2018-06-05T18:24:03.829Z; git: d5fc0523cfa96bfebfbda19606cad384d772f04c; jvm 1.8.0_131-b11
[main] INFO org.eclipse.jetty.server.AbstractConnector - Started ServerConnector@6a192cfe{HTTP/1.1,[http/1.1]}{0.0.0.0:8081}
[main] INFO org.eclipse.jetty.server.Server - Started @1920ms
Entered Servlet: Request(POST //localhost:8081/foo)@35ea7fc6
original status: 200
Entered Servlet: Request(POST //localhost:8081/foo)@35ea7fc6
original status: 500
bar parameter: 123
```

Note the two entered-servlet and original-status lines.

**Comment by joakime** [comment]
Greg is referring to [`HttpServletRequest.getDispatcherType()`](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getDispatcherType--).

Since you are using the `ServletHandler` directly, you'll have to handle things yourself that the `ServletContextHandler` does for you.

First request arrives as `DispatcherType.REQUEST`, is starting to be processed.
That request results in an Exception that flows out.
That error has to be handled by something, so it does the standard error dispatching, aka `DispatcherType.ERROR` to the one and only destination you have declared in your example (the `ServletHandler`).

If you had used `ServletContextHandler` you could do this ...

``` java
package org.eclipse.jetty.demo;

import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.DefaultHandler;
import org.eclipse.jetty.server.handler.HandlerList;
import org.eclipse.jetty.servlet.ErrorPageErrorHandler;
import org.eclipse.jetty.servlet.ServletContextHandler;

public class Demo
{
    public static void main(String[] args) throws Exception
    {
        Server server = new Server(8081);

        ServletContextHandler context = new ServletContextHandler();
        context.setContextPath("/");
        context.addServlet(OldFooServlet.class, "/foo");
        context.addServlet(ErrorServlet.class, "/error");
        ErrorPageErrorHandler errorHandler = new ErrorPageErrorHandler();
        errorHandler.addErrorPage(500, "/error");
        errorHandler.addErrorPage(404, "/error");
        context.setErrorHandler(errorHandler);

        HandlerList handlers = new HandlerList();
        handlers.addHandler(context);
        handlers.addHandler(new DefaultHandler());

        server.setHandler(handlers);

        try
        {
            server.start();
            server.join();
        }
        catch (Throwable t)
        {
            t.printStackTrace();
        }
        finally
        {
            server.stop();
        }
    }

    public static class OldFooServlet extends HttpServlet
    {
        @Override
        protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException
        {
            System.out.println("Entered Servlet: " + req);
            super.service(req, resp);
        }

        @Override
        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            resp.getWriter().println("received get");
        }

        @Override
        protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            System.out.printf("original status: %s%n", resp.getStatus());
            String barValue = req.getParameter("bar");
            System.out.printf("bar parameter: %s%n", barValue);
            resp.getWriter().printf("received post %s", barValue);
        }
    }

    public static class ErrorServlet extends HttpServlet
    {
        @Override
        protected void service(HttpServletRequest req, HttpServletResponse resp) throws IOException
        {
            resp.getWriter().printf("Whoops, we are in an error (DispatcherType=%s)%n", req.getDispatcherType());
        }
    }
}
```

**Client Request**

``` none
$ nc localhost 8081 < badRequest
HTTP/1.1 500 Server Error
Content-Length: 51
Connection: close
Server: Jetty(9.4.11.v20180605)

Whoops, we are in an error (DispatcherType=ERROR)
```

**Comment by joakime** [comment]
If you go this route, know that when you are processing requests with `DispatcherType.ERROR` , there are various request attributes you can use to access the details of the error.

See: https://docs.oracle.com/javaee/7/api/javax/servlet/RequestDispatcher.html

Look at all of the constants starting with `ERROR_` for how to get to that information during your ERROR dispatch.

**Comment by AgentME** [comment]
Ohhh I get it now. ServletHandler runs a new request with dispatcherType=ERROR so I get a chance to show an error page. ServletContextHandler separates that out. I was just caught by surprise by Jetty's error handling because getParameter actually throws an error now unlike in older versions. Thanks for the help!
